### DISCUSSION TITLE
Disable <CR> for completion

### DISCUSSION THREAD

**Discussion by kescobo** [LazyVim/LazyVim#4620] [Q&A] [answered]
Related to #1050 #2885 #2981 #2355, which describe variously disabling the cmp tooltip, toggling autocompletion, etc, changing all the keymaps, etc.

All I want to do is disable the use of `<CR>` to accept the suggestion. Given the [docs here](https://www.lazyvim.org/configuration/plugins#%EF%B8%8F-adding--disabling-plugin-keymaps), I attempted the following in `lua/plugins/cmp.lua`:

```
return {
    "hrsh7th/nvim-cmp",
    -- ...
    keys = {
        { "<CR>", false }
    }
}
```

But this doesn't seem to work. Based [on this page](https://www.lazyvim.org/plugins/coding#nvim-cmp), it seems like these keymaps are set in the `opts` function, if I just paste that  whole opts section and modify it to the following, I get the behavior I want:

```
opts = function()
    -- keep everything from the default
    mapping = cmp.mapping.preset.insert({
      ["<C-b>"] = cmp.mapping.scroll_docs(-4),
      ["<C-f>"] = cmp.mapping.scroll_docs(4),
      ["<C-n>"] = cmp.mapping.select_next_item({ behavior = cmp.SelectBehavior.Insert }),
      ["<C-p>"] = cmp.mapping.select_prev_item({ behavior = cmp.SelectBehavior.Insert }),
      ["<C-Space>"] = cmp.mapping.complete(),
      -- ["<CR>"] = LazyVim.cmp.confirm({ select = auto_select }), -- DEFAULT
      -- ["<C-y>"] = LazyVim.cmp.confirm({ select = true }), -- DEFAULT
      ["<C-y>"] = LazyVim.cmp.confirm({ select = auto-select }), -- ADDED
      ["<S-CR>"] = LazyVim.cmp.confirm({ behavior = cmp.ConfirmBehavior.Replace }), -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
      ["<C-CR>"] = function(fallback)
        cmp.abort()
        fallback()
      end,
    }),
    -- keep everything else from default
end
```

But this seems like there should be a much simpler way to just delete this mapping. I wasn't sure if I could just use a subset of this `opts` function, or if there's a different built-in way. 

Separately, it might be nice to document this somewhere (if it's already documented, maybe make it more discoverable from the adding/disabling plugin keymaps page linked above - I did a fair bit of searching and reading through other discussions before coming here).

**Comment by dpetka2001** [comment, marked as answer]
You can do
```lua
return {
  {
    "hrsh7th/nvim-cmp",
    opts = function(_, opts)
      local cmp = require("cmp")

      opts.mapping = vim.tbl_deep_extend("force", opts.mapping, {
        ["<CR>"] = cmp.config.disable,
      })
    end,
  },
}
```
> Separately, it might be nice to document this somewhere (if it's already documented, maybe make it more discoverable from the adding/disabling plugin keymaps page linked above - I did a fair bit of searching and reading through other discussions before coming here).

Regarding this, I'm just a simple user like you and don't have any say in this project. But what you're asking here is essentially some Lua knowledge. Everything in Lua is a table. `opts` is a table as well with different subtables for each field (like mapping). Tables can be merged together and that's where Neovim API comes in (`vim.tbl_deep_extend` in this case). So, I don't really know how much sense would make to include a section in the LazyVim docs about Lua knowledge.

Like I said, I'm just a simple user like you and I'm just expressing my personal opinion.

**Reply by kescobo** [reply]
This is helpful, thank you! And fair enough, but my point is that there's a section of the docs about enabling / disabling keymaps for packages. `neovim-cmp` is a package, and has keymaps, but it wasn't clear anywhere that I could find suggesting that this package's keymaps are somehow special, why they are special, and how to treat them differently. 

Even something like 

> For some packages, keymaps are defined in `opts`, and must be modified by updating their `opts` table using `vim.tbl_deep_extend`

or something would at least point users in the right direction.
